### ISSUE TITLE
ZStream.fromJavaStream*: Consider always calling close on underlying stream

### ISSUE THREAD

**Issue by felher** [zio/zio#3847]
Hey folks,

I thought it might be useful to always call close when creating a `ZStream` from a java `Stream`. Consider the following code:

```scala
ZStream.fromJavaStream(Files.walk(Paths.get("."))).runHead.forever.orDie
```

As one would expect, after some time it fails with:
```
[ERROR] Fiber failed.
[ERROR] An unchecked error was produced.
[ERROR] java.nio.file.FileSystemException: .: Too many open files
```

I think it might be useful to always call `close` on the underlying stream when the `zstream` is done. Some arguments in favor would be:

1. The stream is created by calling  `iterator()` on it, which is a terminal operation. So most calls on the underlying stream will fail on it anyway, whether it is closed or not.
2. Forgetting to call `close`, which I expect will happen fairly regularly to beginners, leads to very hard to find bugs. Often, they only manifest in production after quite some time. Since the error is caused by exhaustion of some resource, the code which fails in the end might not be the culprit at all.
3. I think quite a few users would actually expect ZIO to handle that, since resource safety is a repeated goal of ZIO (and, I think) ZIO-streams.
4. I expect the performance cost to be negligible

I think it may make sense to introduce a new constructor which doesn't call `close` in case somebody really needs to do it themselves, for example to run the `onClose` handlers. I don't think there is need to provide a `noClose` for all variants, since I expect that one to be used rarely. Maybe `fromJavaStreamManagedNoClose` would suffice?

If you think this is a good idea, I'm happy to open an PR which adds tests, rewrites the old constructors to call `close` and introduce a new constructor which doesn't call close. I'll also update the javadocs to state that `close` is automatically called after the stream is consumed.

If you think this isn't a good idea, feel free to just close it :)

**Comment by iravid** [comment]
Yes this is necessary, was going to open an issue about this myself. A PR would be very welcome!

As for closing the stream or not - let's do the same as `fromInputStream`:
- `fromJavaStream(Stream)` does not close
- `fromJavaStream(ZIO[Stream])` closes
- `fromJavaStream(ZManaged[Stream])` does not close

I think we'll do a little re-arrangement on this scheme before declaring binary compatibility on ZStream but this is ok for now.

Thank you for raising the issue @felher!

**Comment by simpadjo** [comment]
IMHO `Stream`s should be closed much more aggressively than `InputStream`s. 
Once allocated, an  `InputStream` has to be closed. This is a common knowledge in Java.
OTOH `Stream`s are usually autoclosed by terminal operators. 
`fromJavaStream` looks more like a terminal operator for a `Stream`.
I'd either just always close the stream or add an optional `close: Boolean = true` parameter to all relevant methods.

**Comment by felher** [comment]
I, too, would prefer a more aggressive stance on closing when talking about java streams. I feel the same way about it as @simpadjo does: `fromJavaStream` - and its alternatives - look like terminal operations. Not only do they look like it, they are (and have to be).

Here is a somewhat lengthy example:
```scala
import zio._
import zio.stream._
import zio.console._
import java.io._

object Main extends App {
  val ioStream = new ByteArrayInputStream(Array(0: Byte, 1: Byte))
  val stStream = java.util.stream.Stream.of(0: Byte, 1: Byte)

  def run(args: List[String]) = for {
    a1 <- ZStream.fromInputStream(ioStream, 1).runHead.orDie
    _ <- putStrLn(a1.toString)
    a2 <- ZStream.fromInputStream(ioStream, 1).runHead.orDie
    _ <- putStrLn(a2.toString)

    b1 <- ZStream.fromJavaStream(stStream).runHead.orDie
    _ <- putStrLn(b1.toString)
    b2 <- ZStream.fromJavaStream(stStream).runHead.orDie
    _ <- putStrLn(b2.toString)
  } yield ExitCode.success
}
```

I think it reasonable to expect both `a1` and `a2` to work, which they do. And it seems to me that this could be a real use case. I think it is also reasonable to expect `b2` to fail. Currently, the exception you get for `b2` is

> java.lang.IllegalStateException: stream has already been operated upon or closed

which is true, since using the effect provided by `ZStream.fromJavaStream` is a terminal operation. So there is a real difference to how `java.stream.Stream` and `java.io.InputStream` behave, which could justify the asymmetry between both constructor groups. 

On the other hand, I see why symmetry is helpful when designing an API, so I'm happy to implement whatever ends up being.

**Comment by hearnadam** [comment]
Is this still an issue?
